src/main/java/pingis/services/TaskInstanceService.java

```diff
@@ -162,5 +166,22 @@ public TaskInstance getRandomTaskInstance(Task task) {
     return viableInstances.get(new Random().nextInt(viableInstances.size()));
   }
 
+  public TaskInstance getUnfinishedInstance(Challenge challenge, User player) {
+    Optional<TaskInstance> unfinished = getAllByChallenge(challenge).stream()
+        .filter(i -> i.getUser().equals(player))
+        .filter(i -> i.getStatus() == CodeStatus.IN_PROGRESS)
+        .findFirst();
+    if (unfinished.isPresent()) {
+      return unfinished.get();
+    }
+    return null;
+  }
 
-}
+  public boolean canPlayOrSkip(TaskInstance taskInstance) {
+    if (!taskInstance.getUser().equals(userService.getCurrentUser())
+        || taskInstance.getStatus() != CodeStatus.IN_PROGRESS) {
+      return false;
+    }
+    return true;
+  }
```

Comment:
Can just return `taskInstance.getUser().equals(userService.getCurrentUser()) && taskInstance.getStatus() == CodeStatus.IN_PROGRESS` or alternatively make the method a bit longer and make the values into variables, ending with something like `return userOwnsTaskInstance && taskInstanceInProgress`.

Reply:
Result of a careless copypaste. Should have noticed that.